# Introduction
In this project, I will be working with a used car data set from eBay Kleinanzeigen, a classified section of the German eBay website.

## The Dataset
This dataset will have 50,000 data points from the full dataset. This dataset will not be cleaned and I will clean the dataset to prepare it for detailed analysis.

The dataset will have the following columns and defitions

| Column ID | Definition |
| - | - |
| <font color='red'> dateCrawled </font> | When this ad was first crawled. All field-values are taken from this date. |
|<font color='red'>name</font>| Name of car|
|<font color='red'>seller</font>| Whether the seller is private or a deadler  |
|<font color='red'>offerType</font> | The type of listing |
|<font color='red'>price</font> | The price on the ad to sell the car |
|<font color='red'>abtest</font> | Whether the listing is included in an A/B test. |
|<font color='red'>vehicleType</font> | The vehcile type |
|<font color='red'>yearOfRegistration</font> | The year in which the car was first registered. |
|<font color='red'>gearbox</font> | The transmission type |
|<font color='red'>powerPS</font> | The power of the car in PS |
|<font color='red'>model</font> | The car model name |
|<font color='red'>kilometer</font> | How many kilometers the car has driven |
|<font color='red'>monthOfRegistration</font> | The month in which the car was first registered. |
|<font color='red'>fuelType</font> | What type of fuel the car uses. |
|<font color='red'>brand</font> | The brand of the car |
|<font color='red'>notRepairedDamage</font> | If the car has a damage which is not yet repaired |
|<font color='red'>dateCreated</font> | The date on which the eBay listing was created. |
|<font color='red'>nrOfPictures</font> | The number of pictures in the ad. |
|<font color='red'>postalCode</font> | The postal code for the location of the vehicle. |
|<font color='red'>lastSeenOnline</font> | When the crawler saw this ad last online.  |




In [1]:
import pandas as pd
import numpy as np

autos = pd.read_csv('autos.csv' , encoding = 'Windows-1252')

### Note
I initially read this csv with UTF-8 encoding, but it failed to read so I had to try other encodings and it finally succedded with Windows-1252 encoding.

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [4]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

### Note

Taking a look at the autos' dataframe head and info methods, I can see that there are some columns that will need to be converted into number formats so I can make some analysis such as the 'odemeter' column which is currently an "object" type (a string in this scenario).

To make this even more readable, I could convert the german into english

## Next Steps
### Cleaning Column Names
Now that I have the csv file converted into a data frame, I want to convert the column names from camelCase to snake_case and change a few columns shown in the table below for readability.

| Current Col Name | New Name |
| - | - |
|<font color='red'>yearOfRegistration</font> | <font color='red'>registration_year</font> |
|<font color='red'>monthOfRegistration</font> | <font color='red'>registration_month</font> |
|<font color='red'>notRepairedDamage</font> | <font color='red'>unrepaired_damage</font> |
|<font color='red'>dateCreated</font> | <font color='red'>ad_created</font> |

In [5]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
autos_columns = autos.columns
column_fixes = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_PS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']
#if we wanted to replace the columns directly...
#autos.rename(columns = {'dateCrawled' : 'date_crawled', 'offerType' : 'offer_type', 'vehicleType' : 'vehicle_type', 'yearOfRegistration' : 'registration_year', 'powerPS': 'power_PS', 'monthOfRegistration' : 'registration_month', 'fuelType' : 'fuel_type', 'notRepairedDamage' : 'unrepaired_damage', 'dateCreated' : 'ad_created', 'nrOfPictures' : 'nr_of_pictures', 'postalCode' : 'postal_code', 'lastSeen': 'last_seen'})
autos.columns = column_fixes
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_PS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [7]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_PS   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

### Note
Alright! I've successfully renamed the columns into snake_case and renamed the column names to be easily understood.

## Next Steps
### Evaluating Each Column For Usefulness
The first steps I want to take now are looking for text columns where almost or all of the values are the same and dropping them from the dataset because they don't have useful information for the analysis.

In [8]:
autos.describe(include = 'all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-10 15:36:24  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_PS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

### Note
From the descriptive statistics summary table above, we can see that the following columns should be converted from string to a numerical format for a more detailed analysis: price and odometer.

The column that we can drop because they give us no significant insight is: nr_of_pictures, seller and offer_type

The columns that need more investigation are: date_crawled, ad_created and last_seen. These three columns are in date and time formats that can be extracted and cleaned.

## Next Steps
I will now do the following:
* remove any non-numeric characters
* convert the column to a numeric dtype
* rename the odometer column to odeometer_km

In [9]:
#Convert Price to Numeric from text
autos["price"] = autos["price"].str.replace("$","").str.replace(",","").astype(int)


In [10]:
#View the summary statistics of price
autos["price"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [11]:
#Convert odometer to Numeric from text
autos["odometer"] = autos["odometer"].str.replace(",","").str.replace("km","").astype(int)

In [12]:
#Change odometer to odometer_km
autos.rename(columns={"odometer" : "odometer_km"}, inplace = True)


In [13]:
#View the summary statistics of odometer
autos["odometer_km"].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

### Note
Hmm... it's odd that the top 50% of the data has an exact value of 150kms... let's revisit this later and see what values are being represented for the odometer

## Next Steps
I will now explore the odometer_km and price columns to look for anomalies. 

In [14]:
#Explore the price column's shape
print("There are {0} unique values in the price column\n".format(autos["price"].unique().shape[0]))

#Explore the summary statistics of the price column
print("The summary statistics of the price column are:\n", autos["price"].describe(), "\n")

#Counting the frequency of values in the price column in descending order
print("\nThe frequency table of the top 20 values in the price column are:\n", autos["price"].value_counts().sort_index(ascending = False).head(20))

#Counting the frequency of values in the price column in ascending order
print("\nThe frequency table of the bottom 20 values in the price column are:\n", autos["price"].value_counts().sort_index(ascending = True).head(20))

#Counting the frequency of values in the price column in the range 2000 and 100000 are:
print("\nThe frequency table of the values in the price column in range 2000-100000 are:\n", autos.loc[autos["price"].between(2000,100000), "price"].value_counts().sort_index(ascending = True).head(20))

#Removing the outliers
auto_price_adjusted = autos.loc[autos["price"].between(2000,100000), "price"]

#Exploring the data without the outliers
print("\nThere are {0} unique values in the auto price adjusted column\n".format(auto_price_adjusted.unique().shape[0]))
print("The summary statistics of the auto price adjusted column are:\n", auto_price_adjusted.describe(), "\n")



There are 2357 unique values in the price column

The summary statistics of the price column are:
 count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64 


The frequency table of the top 20 values in the price column are:
 99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64

The frequency table of the bottom 20 values in the price column are:
 0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price, dtype: int64

The

In [15]:
#Explore the odometer_km column's shape
print("There are {0} unique values in the odometer_km column\n".format(autos["odometer_km"].unique().shape[0]))

#Explore the summary statistics of the odometer_km column
print("The summary statistics of the odometer_km column are:\n", autos["odometer_km"].describe(), "\n")

#Counting the frequency of values in the odometer_km column in descending order
print("\nThe frequency table of the top 20 values in the odometer_km column are:\n", autos["odometer_km"].value_counts().sort_index(ascending = False).head(20))

#Counting the frequency of values in the odometer_km column in ascending order
print("\nThe frequency table of the bottom 20 values in the odometer_km column are:\n", autos["odometer_km"].value_counts().sort_index(ascending = True).head(20))

#Counting the frequency of values in the odometer_km column in the range 2000 and 200000 are:
print("\nThe frequency table of the values in the price column in range 3000-200000 are:\n", autos.loc[autos["odometer_km"].between(2000,200000), "price"].value_counts().sort_index(ascending = True).head(20))


There are 13 unique values in the odometer_km column

The summary statistics of the odometer_km column are:
 count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64 


The frequency table of the top 20 values in the odometer_km column are:
 150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
20000       784
10000       264
5000        967
Name: odometer_km, dtype: int64

The frequency table of the bottom 20 values in the odometer_km column are:
 5000        967
10000       264
20000       784
30000       789
40000       819
50000      1027
60000      1164
70000      1230
80000      1436
90000      1757
100000     2169
125000     5170
150000    32424
Name: odometer_km, dtype: int64

The frequency table of the values in th

## Next Steps

I'm going to move on to the date columns and clean the format and data such that I can make better analysis with the given data.

There are 5 different date columns: 

* date_crawled
* last_seen
* ad_created
* registration_month
* registration_year

Each of these columns have the following format: YYYY-MM-DD HR:MIN:SEC
I am going to extract the date since that would give me the most value in data analysis. 

I will be focusing on date_crawled, ad_created and last_seen

In [16]:
autos["date_crawled"].str[:10].value_counts(dropna = False, normalize = True).sort_index(ascending = True)

2016-03-05    0.02538
2016-03-06    0.01394
2016-03-07    0.03596
2016-03-08    0.03330
2016-03-09    0.03322
2016-03-10    0.03212
2016-03-11    0.03248
2016-03-12    0.03678
2016-03-13    0.01556
2016-03-14    0.03662
2016-03-15    0.03398
2016-03-16    0.02950
2016-03-17    0.03152
2016-03-18    0.01306
2016-03-19    0.03490
2016-03-20    0.03782
2016-03-21    0.03752
2016-03-22    0.03294
2016-03-23    0.03238
2016-03-24    0.02910
2016-03-25    0.03174
2016-03-26    0.03248
2016-03-27    0.03104
2016-03-28    0.03484
2016-03-29    0.03418
2016-03-30    0.03362
2016-03-31    0.03192
2016-04-01    0.03380
2016-04-02    0.03540
2016-04-03    0.03868
2016-04-04    0.03652
2016-04-05    0.01310
2016-04-06    0.00318
2016-04-07    0.00142
Name: date_crawled, dtype: float64

### Obeservations of date_crawled

The day with the highest frequency of data being crawled was on 2016-04-03 and lowest frequency was 2016-04-07. There isn't much significance that I can draw from at this moment in time, but with further analysis these days may come in handy.

In [17]:
autos["ad_created"].str[:10].value_counts(dropna = False, normalize = True)

2016-04-03    0.03892
2016-03-20    0.03786
2016-03-21    0.03772
2016-04-04    0.03688
2016-03-12    0.03662
2016-03-14    0.03522
2016-04-02    0.03508
2016-03-28    0.03496
2016-03-07    0.03474
2016-03-29    0.03414
2016-03-19    0.03384
2016-04-01    0.03380
2016-03-15    0.03374
2016-03-30    0.03344
2016-03-08    0.03334
2016-03-09    0.03324
2016-03-22    0.03280
2016-03-11    0.03278
2016-03-26    0.03256
2016-03-23    0.03218
2016-03-31    0.03192
2016-03-25    0.03188
2016-03-10    0.03186
2016-03-17    0.03120
2016-03-27    0.03090
2016-03-16    0.03000
2016-03-24    0.02908
2016-03-05    0.02304
2016-03-13    0.01692
2016-03-06    0.01512
               ...   
2016-02-21    0.00006
2016-02-18    0.00004
2016-02-09    0.00004
2016-02-14    0.00004
2016-02-02    0.00004
2016-02-20    0.00004
2016-01-10    0.00004
2016-02-24    0.00004
2016-02-26    0.00004
2016-02-05    0.00004
2015-09-09    0.00002
2016-01-03    0.00002
2015-11-10    0.00002
2016-01-13    0.00002
2016-02-07

### Observations of ad_created

The highest day of ad creation was on 2016-04-03. Is this a coincedence that this is also the highest day that the ad was crawled?

In [18]:
autos["last_seen"].str[:10].value_counts(dropna = False, normalize = True)

2016-04-06    0.22100
2016-04-07    0.13092
2016-04-05    0.12428
2016-03-17    0.02792
2016-04-03    0.02536
2016-04-02    0.02490
2016-03-30    0.02484
2016-04-04    0.02462
2016-03-31    0.02384
2016-03-12    0.02382
2016-04-01    0.02310
2016-03-29    0.02234
2016-03-22    0.02158
2016-03-28    0.02086
2016-03-21    0.02074
2016-03-20    0.02070
2016-03-24    0.01956
2016-03-25    0.01920
2016-03-23    0.01858
2016-03-26    0.01696
2016-03-16    0.01644
2016-03-27    0.01602
2016-03-15    0.01588
2016-03-19    0.01574
2016-03-14    0.01280
2016-03-11    0.01252
2016-03-10    0.01076
2016-03-09    0.00986
2016-03-13    0.00898
2016-03-08    0.00760
2016-03-18    0.00742
2016-03-07    0.00536
2016-03-06    0.00442
2016-03-05    0.00108
Name: last_seen, dtype: float64

### Observations of last_seen

The day with the highest last seen date was 2016-04-06 which is 3 days after the days of the highest ad_creation date. I could possibly explore this further and see if there was any relation to the type of vehicles sold that were on the website for 3 days and make a model to estimate when a specific vehicle would be sold.

In [19]:
autos["registration_year"].describe()

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

### Observations of registration_year

From the summary statistics we can see some obvious outliers here that are going to be labeled as "bad" data. The max and min years are 9999 and 1000  respectively which is highly questionable since it just doesn't fit the appropriate time line that we are scoped to.

Since the first car was built around the 1900s and the date of this data scraping was in 2016, I am going to throw away rows that are outside of these ranges.

In [20]:
#Remove years that are out the 1900-2016 range and assign to new_autos
new_autos = autos[(autos["registration_year"] > 1900) & (autos["registration_year"] <= 2016)]
new_autos["registration_year"].describe()

count    48028.00000
mean      2002.80351
std          7.31085
min       1910.00000
25%       1999.00000
50%       2003.00000
75%       2008.00000
max       2016.00000
Name: registration_year, dtype: float64

In [21]:
new_autos["registration_year"].value_counts(normalize = True).head(10)

2000    0.069834
2005    0.062776
1999    0.062464
2004    0.056988
2003    0.056779
2006    0.056384
2001    0.056280
2002    0.052740
1998    0.051074
2007    0.047972
Name: registration_year, dtype: float64

## Observations after removing outliers

After removing the outliers, the numbers looked to be more aligned with something realistic. The standard deviation was reduced from 105 years to 7 years and the mean car in this dataset was changed to 2002. The top 10 car registration years that were posted the most on the german eBay's listing were ranged from 1998 to 2007 with the highest being cars in 2000.

## Next Steps

Now I want to take this a step further and use aggregation to explore the variations across different car brands. 

In [22]:
#Checking for top 10 brands sold on the german eBay car listing
new_autos["brand"].value_counts().head(10)

volkswagen       10188
bmw               5284
opel              5195
mercedes_benz     4580
audi              4149
ford              3352
renault           2274
peugeot           1418
fiat              1242
seat               873
Name: brand, dtype: int64

### Next steps

I now want to see the oldest registration year by brand. Here I will use aggregation to manipulate data based on each existing unique brand.

In [23]:
oldest_car_by_brand = {}
brands = autos["brand"].unique()

for brand in brands:
    selected_rows = new_autos[new_autos["brand"] == brand].sort_values("registration_year", ascending = True)
    oldest_car_row = selected_rows.iloc[0]
    oldest_car_by_brand[brand] = oldest_car_row["registration_year"]

oldest_car_by_brand


{'alfa_romeo': 1950,
 'audi': 1968,
 'bmw': 1929,
 'chevrolet': 1954,
 'chrysler': 1966,
 'citroen': 1954,
 'dacia': 2005,
 'daewoo': 1996,
 'daihatsu': 1970,
 'fiat': 1960,
 'ford': 1927,
 'honda': 1973,
 'hyundai': 1994,
 'jaguar': 1968,
 'jeep': 1972,
 'kia': 1995,
 'lada': 1976,
 'lancia': 1969,
 'land_rover': 1951,
 'mazda': 1984,
 'mercedes_benz': 1937,
 'mini': 1970,
 'mitsubishi': 1984,
 'nissan': 1981,
 'opel': 1910,
 'peugeot': 1964,
 'porsche': 1950,
 'renault': 1910,
 'rover': 1991,
 'saab': 1983,
 'seat': 1950,
 'skoda': 1958,
 'smart': 1992,
 'sonstige_autos': 1910,
 'subaru': 1988,
 'suzuki': 1980,
 'toyota': 1976,
 'trabant': 1910,
 'volkswagen': 1943,
 'volvo': 1956}

In [24]:
#Get the oldest year of the car listed
min_year = min(oldest_car_by_brand.values())

#Get the brand of the oldest car listed
min_brand = min(oldest_car_by_brand, key = oldest_car_by_brand.get)

print("The minimum year is {0} and the associated brand is {1}".format(min_year,min_brand))

The minimum year is 1910 and the associated brand is trabant


### Observations

One of the oldest car on the german eBay listing site was a 1910 trabant. 

I can expand further on this since I noticed that there are more vehicles with the registration year of 1910. The min function I used only returned the lowest year and the associated brand with that value that it last found. I could next create a list of brands that meet that date, but will do that in the future.

## Next Steps

Next I want to see the mean price of the top 10 brands sold on the website.

In [25]:
#Get top ten brands sold 
top_ten_brands = new_autos["brand"].value_counts().head(10)

#Convert the brands to a list
top_ten_brands_list = top_ten_brands.index.tolist()


In [26]:
#Create dictionary to hold the average price of the top 10 brands
top_10_average_price = {}

for brand in top_ten_brands_list:
    df_containing_brand_only = new_autos[new_autos["brand"] == brand]
    price_sum = df_containing_brand_only["price"].sum()
    how_many_cars_in_brand = df_containing_brand_only.shape[0]
    average_price = price_sum/how_many_cars_in_brand
    top_10_average_price[brand] = average_price
    print("The brand is {0} \nThe price_sum is {1} \nThe number of cars of this brand in this dataset is {2} \nThe average price for this brand is {3}\n ------------------------".format(brand, price_sum, how_many_cars_in_brand, average_price))
top_10_average_price


The brand is volkswagen 
The price_sum is 66389670 
The number of cars of this brand in this dataset is 10188 
The average price for this brand is 6516.457597173145
 ------------------------
The brand is bmw 
The price_sum is 44040265 
The number of cars of this brand in this dataset is 5284 
The average price for this brand is 8334.645155185466
 ------------------------
The brand is opel 
The price_sum is 27287343 
The number of cars of this brand in this dataset is 5195 
The average price for this brand is 5252.61655437921
 ------------------------
The brand is mercedes_benz 
The price_sum is 138853911 
The number of cars of this brand in this dataset is 4580 
The average price for this brand is 30317.447816593885
 ------------------------
The brand is audi 
The price_sum is 37729554 
The number of cars of this brand in this dataset is 4149 
The average price for this brand is 9093.65003615329
 ------------------------
The brand is ford 
The price_sum is 24345629 
The number of cars 

{'audi': 9093.65003615329,
 'bmw': 8334.645155185466,
 'fiat': 2711.8011272141707,
 'ford': 7263.015811455847,
 'mercedes_benz': 30317.447816593885,
 'opel': 5252.61655437921,
 'peugeot': 3039.4682651622,
 'renault': 2395.4164467897976,
 'seat': 4296.492554410081,
 'volkswagen': 6516.457597173145}

In [27]:
highest_average_price = max(top_10_average_price.values())
highest_average_price_brand = max(top_10_average_price, key = top_10_average_price.get)
print("The highest average price is {0} and the associated brand is {1}".format(highest_average_price, highest_average_price_brand))

The highest average price is 30317.447816593885 and the associated brand is mercedes_benz


### Observations

The highest average price is 30317.48 and the associated brand is mercedes benz. This is actually quite high relative to the other top 10 averages. Future investigations would include finding out if there are any price outliers that are causing this to be so high.

I also found that the luxury brands such as audi, bmw and mercedes benz had a higher average price compared to the rest of the top 10.

## Next Steps

Now I want to get the mean milage of the cars and the prices and display them both in one data frame.


In [28]:
#Create dictionary to hold the average milage of the top 10 brands
top_10_average_milage = {}

for brand in top_ten_brands_list:
    df_containing_brand_only = new_autos[new_autos["brand"] == brand]
    milage_sum = df_containing_brand_only["odometer_km"].sum()
    how_many_cars_in_brand = df_containing_brand_only.shape[0]
    average_milage = milage_sum/how_many_cars_in_brand
    top_10_average_milage[brand] = average_milage
    print("The brand is {0} \nThe milage_sum is {1}km \nThe number of cars of this brand in this dataset is {2} \nThe average milage for this brand is {3}km\n ------------------------".format(brand, milage_sum, how_many_cars_in_brand, average_milage))
top_10_average_milage

The brand is volkswagen 
The milage_sum is 1311505000km 
The number of cars of this brand in this dataset is 10188 
The average milage for this brand is 128730.36906164115km
 ------------------------
The brand is bmw 
The milage_sum is 699785000km 
The number of cars of this brand in this dataset is 5284 
The average milage for this brand is 132434.70855412565km
 ------------------------
The brand is opel 
The milage_sum is 671335000km 
The number of cars of this brand in this dataset is 5195 
The average milage for this brand is 129227.14148219442km
 ------------------------
The brand is mercedes_benz 
The milage_sum is 599340000km 
The number of cars of this brand in this dataset is 4580 
The average milage for this brand is 130860.26200873363km
 ------------------------
The brand is audi 
The milage_sum is 536415000km 
The number of cars of this brand in this dataset is 4149 
The average milage for this brand is 129287.78018799711km
 ------------------------
The brand is ford 
The m

{'audi': 129287.78018799711,
 'bmw': 132434.70855412565,
 'fiat': 116553.94524959743,
 'ford': 124046.83770883054,
 'mercedes_benz': 130860.26200873363,
 'opel': 129227.14148219442,
 'peugeot': 127136.81241184767,
 'renault': 128183.81706244503,
 'seat': 121563.57388316152,
 'volkswagen': 128730.36906164115}

In [29]:
mean_milage_series = pd.Series(top_10_average_milage)
mean_price_series = pd.Series(top_10_average_price)

new_mean_df = pd.DataFrame(mean_milage_series, columns=['mean_milage'])

new_mean_df['mean_price'] = mean_price_series

new_mean_df['KM/Dollar'] = new_mean_df['mean_milage'] / new_mean_df['mean_price']

new_mean_df

mean_milage    mean_price  KM/Dollar
audi           129287.780188   9093.650036  14.217369
bmw            132434.708554   8334.645155  15.889664
fiat           116553.945250   2711.801127  42.980270
ford           124046.837709   7263.015811  17.079247
mercedes_benz  130860.262009  30317.447817   4.316335
opel           129227.141482   5252.616554  24.602432
peugeot        127136.812412   3039.468265  41.828636
renault        128183.817062   2395.416447  53.512122
seat           121563.573883   4296.492554  28.293677
volkswagen     128730.369062   6516.457597  19.754655

### Observations

Surprisingly the benz brand has a really high mean milage with the highest mean price. I am curious if there are outliers that are really messing with my data set.

I also added a new column that showed the average milage over the average price. To interpret this column.. anything on the higher end is great bang for your buck such as the renault with a score of 53.51. Anything on the lower end is a really expensive buy for the amount of milage already on that car.. for example the mercedes benz with a score of 4.316.

# Summary & Conclusion

In summary, I cleaned the used car data set from eBay Kleinanzeigen, a classified section of the German eBay website. I found that the mercedes benz brand had the highest average list price (30,317.45) compared to all the other models. I found that the luxury german brands (audi, bmw, mercedes held their resale value compared to other brands (volkswagen, seat, renault..). I also found that one of the oldest vehicles listings was a 1910 trabant.

The skills that I used pandas to clean, explore and understand the car listigs data set.

Some next steps I can do to continue this project is further analyze and aggregate columns to find patterns.